### PR TITLE
[Issue 6024][pulsar_storm] PulsarSpout emit to multiple streams

### LINKED ISSUE #6024: PulsarSpout emits only to default stream
**Is your feature request related to a problem? Please describe.**
We have an application, currently using another streaming ingestion method, that we'd like to replace with either Kafka or Pulsar. Either would be a candidate, _except_ the spout needs to emit to different output streams (to be processed by different bolts) based on the content of individual messages.

The Kafka spout supports this. Topologies implement `RecordTranslator`, an interface that returns the output stream and the `Values` for the tuple.

The equivalent in `PulsarSpout` is `MessageToValuesMapper`, but that returns only the `Values`. There's no way to specify an alternate stream; everything goes to "default."

**Describe the solution you'd like**
Simply put, a way for `MessageToValuesMapper` to specify the output stream for a tuple. (It already has a `declareOutputFields()` method so can declare the appropriate output streams.)

Three possible solutions, all workable. The first avoids any API change but requires an extra check
in `PulsarSpout` on emit. The other two would require backwards-compatible changes to `MessageToValueMapper`. I prefer the first.

1. Define `PulsarTuple` as an extension of `Values` so `MessageToValuesMapper` can return it without changing the API. `PulsarSpout` checks for `PulsarTuple` and emits on the specified stream if found. (_This is the alternative in the pull request._)
1. Change `PulsarSpout` to use a new method in `MessageToValuesMapper` that returns a `PulsarTuple` interface similar to Kafka's `KafkaTuple`. It would return the `Values` _and_ output stream. A default implementation of the new method would just call the existing method (which probably should be deprecated) and return a `PulsarTuple` with stream "default."
1. Add a new method to `MessageToValuesMapper`: `String outputStream(Message msg, Values values)` to return the output stream given the original `Message` and extracted `Values`. The default implementation would just return "default."

**Describe alternatives you've considered**
All of these would require `collector` to be not `private`, and probably other things. None are very satisfying. All would require an application to extend `PulsarSpout`, which can cause compatibility issues with new versions.
1. Change the existing `collector.emit()` calls with a call to a non-private `doEmit(Message msg, Values values)` call. Applications could override that and figure out the output stream as needed. 
1. Make `mapToValueAndEmit()` at least protected so it can be overridden.
1. Override nextTuple() and replicate most of the logic except the `emit()` calls.

Or, just use Kafka, which is what we'd actually do.

**Additional context**
One application receives Avro-encoded events with a header that identifies the schema type and version. (Also some other information not relevant here.) The schema type determines the output stream.

Another application has a fixed Avro schema with two fields, an event type and a serialized protobuf. The event type determines the output stream.

Changing those formats isn't an option.

### PR DESCRIPTION
**Pull request by dawillcox**
Fixes #6024

### Motivation
This is all described in detail in https://github.com/apache/pulsar/issues/6024, but in short, an insurmountable obstacle to using Pulsar in our storm topology is the fact that `PulsarSpout` only emits to the "default" stream. In our environment, we need to emit on different streams based on the content of each received message. This change extends `PulsarSpout` to recognize a `Values` extension that specifies an alternate output stream, and uses that stream when given.

### Modifications
A new `PulsarTuple` class is added. It extends `Values` and adds a method to return the output stream.

When emitting a tuple after calling `toValues(msg)`, `PulsarSpout` checks if the returned `Values` is a `PulsarTuple`. If so, it emits to the designated stream, otherwise it emits as before.

### Verifying this change

- [x] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:
  - A test case was added to `PulsarSpoutTest` to verify that when `PulsarTuple` is used, the tuple is emitted in the specified stream.

### Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (yes?) Existing APIs don't change, but a new class is added that carries additional information.
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: pulsar-storm/src/main/java/org/apache/pulsar/storm/MessageToValuesMapper.java [hunk 1]
```diff
@@ -33,12 +33,12 @@
      * @param msg
      * @return
      */
-    public Values toValues(Message<byte[]> msg);
```
Comment (dawillcox) [inline pulsar-storm/src/main/java/org/apache/pulsar/storm/MessageToValuesMapper.java]: `public` is redundant. Interface messages _must_ be public.

**Comment by dawillcox** [comment]
I'm confused about all of the CI failures. They seem to be mostly related to CPP, which this PR doesn't touch.

File: pulsar-storm/src/main/java/org/apache/pulsar/storm/PulsarTuple.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package org.apache.pulsar.storm;
```
Comment (rdhabalia) [inline pulsar-storm/src/main/java/org/apache/pulsar/storm/PulsarTuple.java]: Please add the license for this file.

**Comment by dawillcox** [comment]
I'll be away until next week. If there's something I need to do to fix these validation failures, let me know. I can't imagine what, though.

**Comment by dawillcox** [comment]
OK, back from time away. @rdhabalia is there something I need to do to move this forward? I want to be able to use this in a topology (or at least make it available for use in topologies), but the Verizon Media guys say it needs to be committed here first, and then they'll pull it in to our local build.

File: pulsar-storm/src/main/java/org/apache/pulsar/storm/PulsarSpout.java [hunk 3]
```diff
@@ -326,7 +326,11 @@ private boolean mapToValueAndEmit(Message<byte[]> msg) {
                 }
                 ack(msg);
             } else {
-                collector.emit(values, msg);
+                if (values instanceof PulsarTuple) {
+                    collector.emit(((PulsarTuple) values).getOutputStream(), values, msg);
+                } else {
+                    collector.emit(values, msg);
+                };
```
Comment (rdhabalia) [inline pulsar-storm/src/main/java/org/apache/pulsar/storm/PulsarSpout.java]: please remove the `;`

**Comment by sijie** [comment]
retest this please

File: pulsar-storm/src/main/java/org/apache/pulsar/storm/PulsarSpout.java (hunk 3 above)
Comment (dawillcox) [reply pulsar-storm/src/main/java/org/apache/pulsar/storm/PulsarSpout.java]: Done.

**Comment by dawillcox** [comment]
As before, several CI tests are failing, but I don't see how this change could cause any failures.

**Comment by sijie** [comment]
@dawillcox there are flaky tests. we have been working on addressing those flaky tests. as for now, I re-run those tests.

**Comment by dawillcox** [comment]
Does it require Godly powers to rerun the tests, or can I kick one off myself? I can't find a button that would let me do that.

**Comment by sijie** [comment]
@dawillcox I rerun some of the failed checks

**Comment by dawillcox** [comment]
Well, we're down to one CI test failure, anyway. Progress.

**Comment by dawillcox** [comment]
Hey hey, everything passed!

**Pull request merged**